### PR TITLE
Add docker file + Add environment variables as method to pass tokens

### PR DESCRIPTION
**Pull request by NotUnlikeTheWaves**
This PR is in two parts:

1. A Dockerfile that allows for a build result (alpine image of ~19MB). To run it, try:
`docker image build -t echo-bot .` followed by `docker run -e ECHO_BOT_TOKEN="Bot <mytoken>" echo-bot`
2. Being able to use the environment token to start up the bot. Running the bot from a token file also still works (though this is currently not supported in the Docker setup)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -45,6 +45,7 @@ token.dat
 .idea/**/workspace.xml
 .idea/**/tasks.xml
 .idea/dictionaries
+.vscode/
```
Comment (UndarkAido) [inline .gitignore:48]: Please use gitignore.io

File: .dockerignore [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Dockerfile
+.dockerignore
+build/
```
Comment (UndarkAido) [inline .dockerignore:3]: Please add at least CLion files - maybe use gitignore.io?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# Use -e ECHO_BOT_TOKEN="Bot exampletoken" as an environment variable when running
+FROM ubuntu:20.04 AS build
+ENV DEBIAN_FRONTEND=noninteractive
+RUN apt update && apt install -y \
+    libboost-all-dev \
+    cmake \
+    build-essential \
+    libssl-dev \
+    libcrypto++-dev \
```
Comment (UndarkAido) [inline Dockerfile:9]: I don't believe libcrypto++ is needed

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    libcurl4-openssl-dev \
+&& rm -rf /var/lib/apt/lists/*
+WORKDIR /echo-bot
```
Comment (UndarkAido) [inline Dockerfile:12]: Could you make this an `echo-bot` configurable so it works if users change the project name? This is a template repository.

File: main.cc [hunk 5]
```diff
@@ -26,22 +27,13 @@ int main(){
...
-	/*/
-	std::string token;
-	{
-		std::ifstream tokenFile("token.dat");
-		if(!tokenFile){
-			std::cerr << "CRITICAL: "
-			          << "There is no valid way for Echo to obtain a token! "
-			          << "Copy the example `token.eg.dat` as `token.dat` to make one.\n";
+	std::string token = getToken();
+	if(token == "") {
```
Comment (UndarkAido) [inline main.cc:31]: use `string::empty` instead

File: main.cc [hunk 6]
```diff
@@ -152,6 +144,31 @@ int main(){
 	return 0;
 }
 
+std::string getToken() {
+	std::string token;
+	
+		/*
+			First attempt to read the token from the ECHO_BOT_TOKEN environment variable.
+		*/
+	char const* env = std::getenv("ECHO_BOT_TOKEN");
```
Comment (UndarkAido) [inline main.cc:153]: Make this more generic? Just `BOT_TOKEN` maybe?

File: Dockerfile (hunk 4 above)
Comment (NotUnlikeTheWaves) [reply Dockerfile:12]: I'm not familiar with template repositories, how do you make this change?

File: Dockerfile (hunk 3 above)
Comment (NotUnlikeTheWaves) [reply Dockerfile:9]: Correct, seems to be the same thing for libcurl4-openssl

File: .dockerignore (hunk 2 above)
Comment (NotUnlikeTheWaves) [reply .dockerignore:3]: I can include the whole .idea folder, but none of this copies over to the final (alpine) image anyway - and I can't think of a big reason to only build the intermediary (ubuntu) image

Comment (NotUnlikeTheWaves) [reply .dockerignore:3]: I have now excluded everything by default except the lib folder and main.cc/cmakelists (i.e. all build necessary kit)

File: Dockerfile (hunk 4 above)
Comment (UndarkAido) [reply Dockerfile:10]: Try `RUN export PROJECT=$(grep project CMakeLists.txt | cut -d "(" -f2 | cut -d ")" -f1)` then use `$PROJECT`

File: main.cc [hunk 7]
```diff
@@ -152,6 +144,31 @@ int main(){
...
 
+std::string getToken() {
+	std::string token;
+	
+		/*
+			First attempt to read the token from the BOT_TOKEN environment variable.
+		*/
+	char const* env = std::getenv("BOT_TOKEN");
+	if(env != nullptr) {
+		token = (std::string) env;
```
Comment (UndarkAido) [inline main.cc:155]: Wouldn't it be safer to use `std::string(env)`?

File: main.cc [hunk 8]
```diff
@@ -152,6 +144,31 @@ int main(){
...
+	} else {
+		/*/
+		 * Read token from token file.
+		 * Tokens are required to communicate with Discord, and hardcoding tokens is a bad idea.
+		 * If your bot is open source, make sure it's ignore by git in your .gitignore file.
+		/*/
+		std::ifstream tokenFile("token.dat");
+		if(!tokenFile){
```
Comment (UndarkAido) [inline main.cc:163]: I'm trying to remember: do you need to close a file even if it fails?

Comment (NotUnlikeTheWaves) [reply main.cc:163]: In this case it doesn't matter because ifstream handles it internally in the destructor

File: main.cc (hunk 7 above)
Comment (NotUnlikeTheWaves) [reply main.cc:155]: I'm not sure it matters much here, env should contain a legitimate series of chars if it pass the null-check

File: Dockerfile (hunk 4 above)
Comment (NotUnlikeTheWaves) [reply Dockerfile:10]: Overall I think this is a nice concept for a template repository, but I think this point is a bit more complicated. To my knowledge the  `RUN export PROJECT=...` only sets it as an environment variable in the intermediary Ubuntu image. You cannot use `WORKDIR /${PROJECT}` in the Dockerfile itself then as this value is not known outside the intermediary Ubuntu image. Another issue is somehow copying this over to the alpine image.

One way around it is to use ARG and define possible arguments for `docker build` but that costs the end user more time to use than actually changing the file.

File: main.cc (hunk 7 above)
Comment (UndarkAido) [reply main.cc:155]: I'd still like this change; it's just considered better form.

**Pull request merged**